src/vm/appdomain.cpp

```diff
@@ -14268,10 +14272,54 @@ HRESULT RuntimeInvokeHostAssemblyResolver(INT_PTR pManagedAssemblyLoadContextToB
                 ObjToArgSlot(_gcRefs.oRefAssemblyName), // AssemblyName instance
             };
 
+            bool fExtractAssemblyFromAssemblyRef = true;
+            
             // Make the call
             _gcRefs.oRefLoadedAssembly = (ASSEMBLYREF) methLoadAssembly.Call_RetOBJECTREF(args);
-            if (_gcRefs.oRefLoadedAssembly != NULL)
+            if (_gcRefs.oRefLoadedAssembly == NULL)
+            {
+                // Step 3 (of CLRPrivBinderAssemblyLoadContext::BindUsingAssemblyName)
+                //
+                // If we could not resolve the assembly using Load method, then bind using TPA binder if present in its
+                // load context.
+                if (pTPABinder != NULL)
+                {
+                    BINDER_SPACE::Assembly *pCoreCLRFoundAssembly = NULL;
+                    hr = pTPABinder->BindAssemblyByNameWorker(pAssemblyName, &pCoreCLRFoundAssembly, true /* excludeAppPaths */);
+                    if (SUCCEEDED(hr))
+                    {
+                        pAssemblyBindingContext = pCoreCLRFoundAssembly;
+                        fExtractAssemblyFromAssemblyRef = false;
+                    }
+                }
+                
+                if (fExtractAssemblyFromAssemblyRef == true) 
```

Comment:
The `fExtractAssemblyFromAssemblyRef` is confusing here. I think it would help if the overall structure of this method would be:

```
   {
      ... Step 1
   }

   if (!fLoaded)
   {
      ... Step 2
   }

   if (!fLoaded)
   {
      ... Step 3
   }

   if (!fLoaded)
   {
      ... Step 4
   }

   ...
```